libhoth_spi.c

```diff
@@ -0,0 +1,286 @@
+
+int libhoth_spi_receive_response(struct libhoth_device* dev, void* response,
+                                 size_t max_response_size, size_t* actual_size,
+                                 int timeout_ms);
+
+int libhoth_spi_close(struct libhoth_device* dev);
+
+
+static int spi_nor_write(int fd, unsigned int address_mode_4b, unsigned int address, const void *data, size_t data_len)
+{
+    if(fd < 0 || !data || !data_len)
+        return LIBHOTH_ERR_INVALID_PARAMETER;
+
+    uint8_t wp_buf[1];
+    uint8_t rq_buf[5];
+    struct spi_ioc_transfer xfer[3];
+    memset(xfer, 0, sizeof xfer);
+    memset(wp_buf, 0, sizeof wp_buf);
+    memset(rq_buf, 0, sizeof rq_buf);
+
+
+    // Write Enable Message
+    wp_buf[0] = 0x06;
+    xfer[0].tx_buf = (unsigned long)wp_buf;
+    xfer[0].len = 1;
+    xfer[0].cs_change = 1;
+
+    // Page Program OPCODE + Mailbox Address
+    rq_buf[0] = 0x02;
+    if(address_mode_4b)
+    {
+        rq_buf[1] = (address >> 24) & 0xFF;
+        rq_buf[2] = (address >> 16) & 0xFF;
+        rq_buf[3] = (address >> 8) & 0xFF;
+        rq_buf[4] = address & 0xFF;
+    
+        xfer[1].len = 5;
+    }
+    else
+    {
+        rq_buf[1] = (address >> 16) & 0xFF;
+        rq_buf[2] = (address >> 8) & 0xFF;
+        rq_buf[3] = address & 0xFF;
+
+        xfer[1].len = 4;
+    }
+    xfer[1].tx_buf = (unsigned long)rq_buf;
+    
+
+    // Write Data at mailbox address
+    xfer[2].tx_buf = (unsigned long)data;
+    xfer[2].len = data_len;
+
+    int status = ioctl(fd, SPI_IOC_MESSAGE(3), xfer);
+    if (status < 0) {
+        return LIBHOTH_ERR_FAIL;
+    }
+
+    return LIBHOTH_OK;
+}
+
+static int spi_nor_read(int fd, unsigned int address_mode_4b, unsigned int address, void *data, size_t data_len)
+{
+    if(fd < 0 || !data || !data_len)
+        return LIBHOTH_ERR_INVALID_PARAMETER;
+
+    uint8_t rd_request[5];
+    struct spi_ioc_transfer xfer[2];
+    memset(xfer, 0, sizeof xfer);
+
+    // Read OPCODE and mailbox address
+    rd_request[0] = 0x03; // Read
+    if(address_mode_4b)
+    {
+        rd_request[1] = (address >> 24) & 0xFF;
+        rd_request[2] = (address >> 16) & 0xFF;
+        rd_request[3] = (address >> 8) & 0xFF;
+        rd_request[4] = address & 0xFF;
+    
+        xfer[0].len = 5;
+    }
+    else
+    {
+        rd_request[1] = (address >> 16) & 0xFF;
+        rd_request[2] = (address >> 8) & 0xFF;
+        rd_request[3] = address & 0xFF;
+
+        xfer[0].len = 4;
+    }
+    xfer[0].tx_buf = (unsigned long)rd_request;
+    
+    // Read in data
+    xfer[1].rx_buf = (unsigned long)data;
+    xfer[1].len = data_len;
+
+    int status = ioctl(fd, SPI_IOC_MESSAGE(2), xfer);
+    if (status < 0) {
+        return LIBHOTH_ERR_FAIL;
+    }
+
+    return LIBHOTH_OK;   
+}
+
+
+int libhoth_spi_open(const struct libhoth_spi_device_init_options* options,
+                     struct libhoth_device** out) {
+  if (out == NULL || options == NULL || options->path == NULL) {
```

Comment:
Something is wonky with the indentation here

Reply:
Fixed indentation